### PR TITLE
Selector: Defend against oldIE cloning caches

### PR DESCRIPTION
**Pull request by gibson042**
This resolves jquery/jquery#1709 by tolerating attroperty cloning, requiring no jQuery changes and reverting gh-310.

@jquery/core & @markelog: thoughts?

### CODE REVIEW

**Comment by timmywil** [comment]
@gibson042 Sometimes I wonder if this cache code is more than its worth. Have we ever done perf testing focused on how much we gain with dir caching? If we could do away with all of this, we'd resolve the issue and have cleaner code.

**Comment by gibson042** [comment]
I think we have checked, but I can't find anything about it now. As I remember, though, it was highly valuable in precisely the cases one would expect: long-prefix selectors against deep DOM trees (since dir caching shortcuts everything to the left of a combinator when a node is visited multiple times by the same filter).

File: dist/sizzle.js [hunk 1]
```diff
@@ -1172,11 +1173,21 @@ Expr = Sizzle.selectors = {
 
 						// non-xml :nth-child(...) stores cache data on `parent`
 						if ( forward && useCache ) {
+
 							// Seek `elem` from a previously-cached index
-							outerCache = parent[ expando ] || (parent[ expando ] = {});
-							cache = outerCache[ type ] || [];
-							nodeIndex = cache[0] === dirruns && cache[1];
-							diff = cache[0] === dirruns && cache[2];
+
```
Comment (markelog) [inline dist/sizzle.js]: Needless blank line?

**Comment by markelog** [comment]
So to provide other side of the argument, that adds a lot of logic and code, whereas if we would just expose the expando and remove it in core code it would be simple without additional byte hit.

This way is more correct, but what we would solve here?

**Comment by gibson042** [comment]
The size difference is negligible (+7 to Sizzle for exposing expando and an additional +7 to jQuery for using it vs. +12 to Sizzle for being inherently cloneNode-safe in oldIE), and [`:nth-last-child(…)` caching](https://github.com/gibson042/sizzle/blob/28f4c67c508719cc76e2075dce0e4cbe0e325919/src/sizzle.js#L1194-L1196) is actually _[clarified](https://github.com/gibson042/sizzle/blob/jquery_gh1709/src/sizzle.js#L1206-L1220)_ by this change (which is essentially a single new lookup/assignment that appears in three places).

The cache data was always supposed to have single-element scope; this just guarantees that cloneNode doesn't violate that intention.

**Comment by markelog** [comment]
Okay then, i assume we don't need additional tests in core? If so, we can close https://github.com/jquery/jquery/issues/1709 from the latest commit in here?

**Comment by gibson042** [comment]
Correct.

**Comment by markelog** [comment]
So you would need to change commit message of the last commit?

**Comment by gibson042** [comment]
I don't think so. What did you have in mind?

**Comment by markelog** [comment]
You know, little stuff, no big deals :-).

We have three commits here with identical commit headers, i would either distinguish them more clearly or unite those changes in one commit. 

And not `Ref jquery/jquery#1709` but `Fixes jquery/jquery#1709`, so we could close that ticket while closing this pull.

Also, https://github.com/jquery/sizzle/pull/328/files#r27660499. But that would be fixed either way with jscs at some point, oh and btw, how about we do https://github.com/jquery/jquery/issues/2056 for the sizzle, before releasing the new version? Since code-styles now is changed so much?

File: dist/sizzle.js (hunk 1 above)
Comment (gibson042) [inline dist/sizzle.js]: It's intentional. The first comment applies broadly, the second only to use of `node` instead of `parent`.

**Comment by gibson042** [comment]
> We have three commits here with identical commit headers, i would either distinguish them more clearly or unite those changes in one commit.

I'll be collapsing them before a merge to master.

> And not `Ref jquery/jquery#1709` but `Fixes jquery/jquery#1709`, so we could close that ticket while closing this pull.

I can change it to "Fixes", but jquery/jquery#1709 won't auto-close—and rightly so, since it still exists until jQuery pulls in the change. @jquery/core: thoughts on "Ref" vs. "Fixes" here?

As for updating grunt-jscs before release, I think that's feasible.

**Comment by dmethvin** [comment]
I'd go with Ref, since as you say it won't be closed until the jQuery update. That update can have `Closes` in its commit message, right? It would be a good practice to do that anyway so we know what we're fixing.

**Comment by markelog** [comment]
> I can change it to "Fixes", but jquery/jquery#1709 won't auto-close—and rightly so, since it still exists until jQuery pulls in the change. @jquery/core: thoughts on "Ref" vs. "Fixes" here?

It should - https://github.com/blog/1439-closing-issues-across-repositories

> since it still exists until jQuery pulls in the change

Right, but the issue itself is fixed, just in another repo, we could do that when we will update Sizzle, but it has good probability of ignoring this issue or issues like this.

If we would follow this logic, we could close jquery tickets (or issues of any other project) only after we would release it.

**Comment by mzgol** [comment]
> Right, but the issue itself is fixed, just in another repo

The issue is fixed if it's fixed on `master` (or `compat` for jQuery Compat). Until Sizzle is updated there, it's therefore not fixed. This is important; otherwise we could release thinking we have no blockers and forget about updating Sizzle first. This has already happened; we have guards in jquery-release but it's better to know about such things before deciding to release.

**Comment by markelog** [comment]
I guess better way would be to make it a sizzle issue, not a core one

**Pull request closed without merging**